### PR TITLE
[FLINK-27758][tests][table] Migrate flink-table-runtime to JUnit5

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

Update the flink-table-runtime module to AssertJ and JUnit 5 following the [JUnit 5 Migration Guide](https://docs.google.com/document/d/1514Wa_aNB9bJUen4xm5uiuXOooOJTtXqS_Jqk9KJitU/edit)

## Brief change log

Use JUnit5 and AssertJ in tests instead of JUnit4 and Hamcrest

## Verifying this change

This change is a code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? ( no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by snuyanzin** [comment]
@flinkbot run azure

**Comment by snuyanzin** [comment]
The PR is based on commit https://github.com/apache/flink/pull/19780  migrating  `TypeSerializerUpgradeTestBase` to JUnit5

**Comment by snuyanzin** [comment]
@Jiabao-Sun since you are one of those who is active in junit4-junit5 migration
may I ask you to have a look here?

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/rank/window/WindowRankOperatorTest.java [hunk 1]
```diff
@@ -127,7 +128,7 @@ public WindowRankOperatorTest(ZoneId shiftTimeZone) {
         this.shiftTimeZone = shiftTimeZone;
     }
 
-    @Parameterized.Parameters(name = "TimeZone = {0}")
+    @Parameters(name = "TimeZone = {0}")
     public static Collection<Object[]> runMode() {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/rank/window/WindowRankOperatorTest.java:132]: NIT: can be private

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/sort/BinaryExternalSorterTest.java [hunk 2]
```diff
@@ -72,7 +73,7 @@ public BinaryExternalSorterTest(boolean spillCompress, boolean asyncMerge) {
         }
     }
 
-    @Parameterized.Parameters(name = "spillCompress-{0} asyncMerge-{1}")
+    @Parameters(name = "spillCompress-{0} asyncMerge-{1}")
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/sort/BinaryExternalSorterTest.java:76]: NIT: can be private

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/window/tvf/operator/UnalignedWindowTableFunctionOperatorTest.java [hunk 3]
```diff
@@ -48,15 +49,20 @@
...
 
-    public UnalignedWindowTableFunctionOperatorTest(ZoneId shiftTimeZone) {
+    UnalignedWindowTableFunctionOperatorTest(ZoneId shiftTimeZone) {
         super(shiftTimeZone);
     }
 
-    @Test
-    public void testEventTimeSessionWindows() throws Exception {
+    @Parameters(name = "TimeZone = {0}")
+    public static Collection<Object[]> runMode() {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/window/tvf/operator/UnalignedWindowTableFunctionOperatorTest.java:60]: NIT: can be private

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/window/tvf/slicing/WindowedSliceAssignerTest.java [hunk 4]
```diff
@@ -31,22 +33,22 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 /** Tests for {@link SliceAssigners.WindowedSliceAssigner}. */
-@RunWith(Parameterized.class)
-public class WindowedSliceAssignerTest extends SliceAssignerTestBase {
+@ExtendWith(ParameterizedTestExtension.class)
+class WindowedSliceAssignerTest extends SliceAssignerTestBase {
 
-    @Parameterized.Parameter public ZoneId shiftTimeZone;
+    public ZoneId shiftTimeZone;
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/window/tvf/slicing/WindowedSliceAssignerTest.java:39]:
```suggestion
    private ZoneId shiftTimeZone;
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixerTest.java [hunk 5]
```diff
@@ -114,10 +111,9 @@ public static List<TestSpec> testData() {
                                                                 .bridgedTo(Time.class))))));
     }
 
-    @Parameterized.Parameter public TestSpec testSpec;
-
-    @Test
-    public void testPrecisionFixing() {
+    @ParameterizedTest(name = "{index}: [From: {0}, To: {1}]")
+    @MethodSource("testData")
+    public void testPrecisionFixing(final TestSpec testSpec) {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixerTest.java:116]:
```suggestion
   void testPrecisionFixing(final TestSpec testSpec) {
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixerTest.java [hunk 6]
```diff
@@ -49,10 +48,8 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 /** Tests for {@link DataTypePrecisionFixer}. */
-@RunWith(Parameterized.class)
 public class DataTypePrecisionFixerTest {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixerTest.java:51]:
```suggestion
class DataTypePrecisionFixerTest {
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixerTest.java [hunk 7]
```diff
@@ -49,10 +48,8 @@
...
 
-    @Parameterized.Parameters(name = "{index}: [From: {0}, To: {1}]")
     public static List<TestSpec> testData() {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/DataTypePrecisionFixerTest.java:53]:
```suggestion
    private static List<TestSpec> testData() {
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/LogicalTypeAssignableTest.java [hunk 8]
```diff
@@ -63,10 +60,8 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 /** Tests for {@link PlannerTypeUtils#isAssignable(LogicalType, LogicalType)}. */
-@RunWith(Parameterized.class)
 public class LogicalTypeAssignableTest {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/LogicalTypeAssignableTest.java:63]:
```suggestion
class LogicalTypeAssignableTest {
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/LogicalTypeAssignableTest.java [hunk 9]
```diff
@@ -63,10 +60,8 @@
...
 
-    @Parameters(name = "{index}: [{0} COMPATIBLE {1} => {2}")
     public static List<Object[]> testData() {
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/LogicalTypeAssignableTest.java:65]:
```suggestion
    private static List<Object[]> testData() {
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/LogicalTypeAssignableTest.java [hunk 10]
```diff
@@ -206,16 +201,10 @@ public static List<Object[]> testData() {
...
-    public LogicalType targetType;
-
-    @Parameter(2)
-    public boolean equals;
-
-    @Test
-    public void testAreTypesCompatible() {
+    @ParameterizedTest(name = "{index}: [{0} COMPATIBLE {1} => {2}")
+    @MethodSource("testData")
+    public void testAreTypesCompatible(
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/types/LogicalTypeAssignableTest.java:206]:
```suggestion
   void testAreTypesCompatible(
```

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/util/collections/binary/BytesHashMapTestBase.java [hunk 11]
```diff
@@ -37,7 +37,7 @@
 import org.apache.flink.table.types.logical.SmallIntType;
 import org.apache.flink.table.types.logical.VarCharType;
 
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
```
Comment (Jiabao-Sun) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/util/collections/binary/BytesHashMapTestBase.java:40]: `public abstract class BytesHashMapTestBase` can be package private.

**Review by Jiabao-Sun — COMMENTED** [review]
Thanks @snuyanzin for the great work!
I left some comments.

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/hashtable/BinaryHashTableTest.java [hunk 12]
```diff
@@ -56,7 +57,7 @@
 import static org.assertj.core.api.Assertions.fail;
 
 /** Hash table it case for binary row. */
-@RunWith(Parameterized.class)
+@ExtendWith(ParameterizedTestExtension.class)
 public class BinaryHashTableTest {
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/hashtable/BinaryHashTableTest.java]: this is actually a good idea, thanks
moved

**Review by Jiabao-Sun — APPROVED** [review]
Thanks @snuyanzin for the quick update. 
LGTM.

**Comment by snuyanzin** [comment]
thanks a lot for taking a look

**Pull request merged**